### PR TITLE
8332507: compilation result depends on compilation order

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Wildcards and separate compilation haven't been friends since like: forever. Javac has had a long standing issue of not being able to restore wildcard bounds when they are read from the classfile. Part of this is due to a restriction of not reading another class file while a given class file is being read. Without breaking this restriction we can "wait" until a class symbol is completed from a class file to have access to richer information and being able to set the right bounds to wildcards. The fix seems simple enough. Basically the proposal is to register listeners that will be invoked once a given class symbol is completely read from a class file. This tactic has been used in type inference and other areas of the compiler where another layer of indirection has been a must to fix similar problems.

It will be a great pleasure if this proposal can be the solution to this long standing issue,

TIA

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8332507: compilation result depends on compilation order`

### Issue
 * [JDK-8332507](https://bugs.openjdk.org/browse/JDK-8332507): compilation result depends on compilation order (**Bug** - P4)(⚠️ The fixVersion in this issue is [24] but the fixVersion in .jcheck/conf is 23, a new backport will be created when this pr is integrated.)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [ba4d99fe](https://git.openjdk.org/jdk/pull/19400/files/ba4d99fe2ad877a6dcf01a6d637fb148b1a5a872)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19400/head:pull/19400` \
`$ git checkout pull/19400`

Update a local copy of the PR: \
`$ git checkout pull/19400` \
`$ git pull https://git.openjdk.org/jdk.git pull/19400/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19400`

View PR using the GUI difftool: \
`$ git pr show -t 19400`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19400.diff">https://git.openjdk.org/jdk/pull/19400.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19400#issuecomment-2130186934)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java [hunk 1]
```diff
@@ -590,7 +590,21 @@ Type classSigToType() {
...
                                                          startSbp,
                                                          sbp - startSbp));
-                outer = new ClassType(outer, sigToTypes('>'), t) {
+                List<Type> actuals = sigToTypes('>');
+                List<Type> formals = ((ClassType)t.type.tsym.type).typarams_field;
+                if (formals != null) {
+                    if (actuals.isEmpty())
+                        actuals = formals;
+                    else
+                        updateBounds(actuals, formals);
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java:599]: Is it important we do this here - or can we always do this in the completer (to simplify the code some more)

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java:599]: thanks for the suggestion, will give it a try

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java [hunk 2]
```diff
@@ -621,7 +633,24 @@ public Type getEnclosingType() {
                         public void setEnclosingType(Type outer) {
                             throw new UnsupportedOperationException();
                         }
-                    };
+
+                        @Override
+                        public List<Type> getTypeArguments() {
+                            List<Type> formalsCp = ((ClassType)t.type.tsym.type).typarams_field;
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java:639]: This works. You might use another boolean flag up top to say whether you actually need to adjust the type arguments (which is only needed the first time) ?

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java]: sure sorry I forgot about that

**Pull request closed without merging**